File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/internal/lifecycle/HandleState.java
package org.maplibre.nativeffi.internal.lifecycle;

import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import org.maplibre.nativeffi.internal.memory.MemoryUtil;
import org.maplibre.nativeffi.internal.status.Status;

/** Shared released-state bookkeeping for native handles. */
public final class HandleState {
  private static final Cleaner CLEANER = Cleaner.create();

  private final String typeName;
  private final MemorySegment handle;
  private final LeakReport leakReport;
  private final Cleaner.Cleanable cleanable;

  @SuppressWarnings("unused")
  private final Object[] parents;

  private boolean released;
  private boolean releasing;
  private int liveChildren;
  private final Map<String, Integer> liveChildCounts = new LinkedHashMap<>();

  public HandleState(String typeName, MemorySegment handle, Object... parents) {
    this.typeName = Objects.requireNonNull(typeName, "typeName");
    this.handle = Objects.requireNonNull(handle, "handle");
    if (MemoryUtil.isNull(handle)) {
      throw new IllegalArgumentException(typeName + " native handle is null");
    }
    this.parents = parents == null ? new Object[0] : parents.clone();
    this.leakReport = new LeakReport(typeName, handle.address());
    this.cleanable = CLEANER.register(this, leakReport);
  }

  public synchronized MemorySegment requireLive() {
    if (released) {
      throw Status.released(typeName);
    }
    if (releasing) {
      throw Status.releasing(typeName);
    }
    return handle;
  }

  public synchronized boolean isReleased() {
    return released;
  }

  public long address() {
    return handle.address();
  }

  public synchronized ChildRetention retainChild(String childTypeName) {
    if (released) {
      throw Status.released(typeName);
    }
    if (releasing) {
      throw Status.releasing(typeName);
    }
    liveChildren++;
    liveChildCounts.merge(childTypeName, 1, Integer::sum);
    return new ChildRetention(this, childTypeName);
  }

  public void closeOnce(NativeDestroy destroy) {
    closeOnce(destroy, () -> {});
  }

  public void closeOnce(NativeDestroy destroy, Runnable afterSuccess) {
    Objects.requireNonNull(destroy, "destroy");
    Objects.requireNonNull(afterSuccess, "afterSuccess");

    synchronized (this) {
      if (released) {
        return;
      }
      if (releasing) {
        throw Status.releasing(typeName);
      }
      if (liveChildren > 0) {
        throw Status.liveChildren(typeName, liveChildren, liveChildSummary());
      }
      releasing = true;
    }

    var destroySucceeded = false;
    try {
      Status.check(destroy.destroy(handle));
      destroySucceeded = true;
    } finally {
      if (!destroySucceeded) {
        synchronized (this) {
          releasing = false;
        }
      }
    }

    synchronized (this) {
      released = true;
      releasing = false;
      leakReport.markReleased();
    }
    cleanable.clean();

    afterSuccess.run();
  }

  private synchronized String liveChildSummary() {
    return String.join(
        ", ",
        liveChildCounts.entrySet().stream()
            .map(entry -> entry.getKey() + "=" + entry.getValue())
            .toList());
  }

  private synchronized void releaseChild(String childTypeName) {
    if (liveChildren > 0) {
      liveChildren--;
    }
    liveChildCounts.computeIfPresent(
        childTypeName, (ignored, count) -> count > 1 ? count - 1 : null);
  }

  @FunctionalInterface
  public interface NativeDestroy {
    int destroy(MemorySegment handle);
  }

  public static final class ChildRetention implements AutoCloseable {
    private final HandleState parent;
    private final String childTypeName;

    private boolean released;

    private ChildRetention(HandleState parent, String childTypeName) {
      this.parent = parent;
      this.childTypeName = Objects.requireNonNull(childTypeName, "childTypeName");
    }

    @Override
    public synchronized void close() {
      if (released) {
        return;
      }
      released = true;
      parent.releaseChild(childTypeName);
    }
  }

  private static final class LeakReport implements Runnable {
    private final String typeName;
    private final long address;
    private volatile boolean released;

    private LeakReport(String typeName, long address) {
      this.typeName = typeName;
      this.address = address;
    }

    private void markReleased() {
      released = true;
    }

    @Override
    public void run() {
      if (!released) {
        System.err.printf(
            "Leaked %s native handle 0x%x; close handles explicitly on their owner thread.%n",
            typeName, address);
      }
    }
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/internal/lifecycle/HandleState.kt
package org.maplibre.nativeffi.internal.lifecycle

import kotlin.concurrent.atomics.AtomicInt
import kotlin.concurrent.atomics.ExperimentalAtomicApi
import kotlin.experimental.ExperimentalNativeApi
import kotlin.native.ref.Cleaner
import kotlin.native.ref.createCleaner
import kotlinx.cinterop.CPointed
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.rawValue
import kotlinx.cinterop.toLong
import org.maplibre.nativeffi.internal.status.Status

/** Shared closed-state bookkeeping for native handles. */
@OptIn(ExperimentalForeignApi::class, ExperimentalAtomicApi::class, ExperimentalNativeApi::class)
internal class HandleState<T : CPointed>(
  private val typeName: String,
  handle: CPointer<T>?,
  vararg parents: Any,
) {
  @Suppress("unused") private val parents: Array<out Any> = parents
  private val address =
    requireNotNull(handle) { "$typeName native handle is null" }.rawValue.toLong()
  private val leakReport = LeakReport(typeName, address)
  @Suppress("unused") private val cleaner: Cleaner = createCleaner(leakReport) { it.report() }
  private val releaseState = AtomicInt(STATE_LIVE)
  private var handle: CPointer<T>? = handle

  fun requireLive(): CPointer<T> =
    when (releaseState.load()) {
      STATE_LIVE -> handle ?: throw Status.released(typeName)
      STATE_RELEASING -> throw Status.invalidState("$typeName is currently releasing")
      else -> throw Status.released(typeName)
    }

  fun isReleased(): Boolean = releaseState.load() == STATE_CLOSED

  fun address(): Long = address

  fun closeOnce(destroy: (CPointer<T>) -> Int) {
    closeOnce(destroy) {}
  }

  fun closeOnce(destroy: (CPointer<T>) -> Int, afterSuccess: () -> Unit) {
    if (!releaseState.compareAndSet(STATE_LIVE, STATE_RELEASING)) {
      when (releaseState.load()) {
        STATE_CLOSED -> return
        STATE_RELEASING -> throw Status.invalidState("$typeName is currently releasing")
        else -> throw Status.released(typeName)
      }
    }
    val live =
      handle
        ?: run {
          releaseState.store(STATE_CLOSED)
          return
        }
    try {
      Status.check(destroy(live))
    } catch (error: Throwable) {
      releaseState.store(STATE_LIVE)
      throw error
    }
    handle = null
    leakReport.markReleased()
    releaseState.store(STATE_CLOSED)
    afterSuccess()
  }

  internal class LeakReport(
    private val typeName: String,
    private val address: Long,
    private val writeLine: (String) -> Unit = { message -> println(message) },
  ) {
    private val released = AtomicInt(0)

    fun markReleased() {
      released.store(1)
    }

    fun report() {
      if (released.load() == 0) {
        writeLine(
          "Leaked $typeName native handle 0x${address.toString(16)}; " +
            "close handles explicitly on their owner thread."
        )
      }
    }
  }

  private companion object {
    private const val STATE_LIVE = 0
    private const val STATE_RELEASING = 1
    private const val STATE_CLOSED = 2
  }
}
